### ISSUE TITLE
Angular 8: Can't resolve all parameters for Component: (?).

### ISSUE THREAD

**Issue by jfcere** [thymikee/jest-preset-angular#288]
I recently upgraded my application to Angular 8 and couldn't get the tests to run successfully with _Jest-Preset-Angular_ resulting in the error below telling me it cannot resolve the constructor dependencies although it works fine using _Karma_ and was working with Angular 7.

```
Can't resolve all parameters for AppComponent: (?).
```

I've created a newly generated Angular 8 application to see if it has something to do with my migration and I ended up having the same issue.

Here is the reproduction repository: https://github.com/jfcere/jest-preset-angular-issue

**Comment by wtho** [comment]
### TL;DR
```diff
tsconfig.spec.json

   "compilerOptions": {
+    "emitDecoratorMetadata": true,
     "outDir": "./out-tsc/spec",

```

### Issue
The Angular CLI takes over some of the TS -> JS transpilation using transformers, especially when it's about the decorators and reflection. We were not aware of this change, but it appearently came in in Angular v8.1.

Here the full PR: https://github.com/angular/angular-cli/pull/14473

We can try to reuse their transformer (https://github.com/angular/angular-cli/blob/master/packages/ngtools/webpack/src/transformers/ctor-parameters.ts), as `@angular/cli` is anyway a dependency in every Angular project.

I will try to experiment with the transformer, we can track the progress in this issue till then.

### Workaround
To make `ts-jest` not lose the parameter information upon compilation without the AST transformation, you can set `"emitDecoratorMetadata": true` in your `tsconfig.spec.json` for now. The PR mentions issues with AOT compilation, which is usually not used in testing, so this setting should not create any problems. If you use karma as well and this leads to issues, try to use a separate `tsconfig.spec.json`.

**Comment by jfcere** [comment]
@wtho thanks for the quick answer, really appreciated!

I've downgraded the reproduction repository to use Angular 8.0.3 and the error is still happening.

But the workaround for `tsconfig.spec.ts` you've provided works! 😄

**Comment by wtho** [comment]
Ok interesting, I created a project with 8.0.0 once and there `emitDecoratorMetadata` was still set, I thought it would not have come in in a patch version update, but for Angular it is just an internal detail.

It's difficult to spot, as I could not find the PR referenced in the [angular-cli releases list](https://github.com/angular/angular-cli/releases) anywhere.

**Comment by jfcere** [comment]
>  I created a project with 8.0.0 once and there emitDecoratorMetadata was still set

Oh alright, that could be it, I've only downgraded the angular dependencies manually in `package.json`, I didn't regenerate a new project with @angular/cli v8.0.x

**Comment by wtho** [comment]
So I was able to wrap the transformer, but I am not sure if this is the way to with `ts-jest`.

### Why?
This example leads to a runtime error in JIT-compiled Angular projects <v7 using `emitDecoratorMetadata`, but in Angular projects v8 it works:
```ts
@Injectable()
export class BService {
  constructor(@Inject(forwardRef(() => AService)) public as: AService) {}
}

@Injectable()
export class AService {
}
```
Runtime Error with `emitDecoratorMetadata`: `ReferenceError: Cannot access 'AService' before initialization`.

More to read about it: https://github.com/angular/angular/issues/30106 and https://github.com/microsoft/TypeScript/issues/27519

Angular AOT always handled this, but JIT only since recently (since this PR https://github.com/angular/angular-cli/pull/14473). Basically a test with this scenario would pass in Angular or karma, but not with our preset. This is achieved using the [DecoratorDownlevelTransformer](https://github.com/angular/angular-cli/blob/master/packages/ngtools/webpack/src/transformers/ctor-parameters.ts).

### What does the transformer do

The transformer stores the constructor parameter types as JS values on the object inside a `ctorParameters`-property:
```
export class SomeClass { constructor(v: ClassInject) {} }

  becomes

export class SomeClass { constructor(v) { } } SomeClass.ctorParameters = () => [ { type: ClassInject } ]
```

### Discussion
To make the existing `angular-cli` transformer work I had to import `typescript` on my own and call `ts.createProgram`, which seems something `ts-jest` tries to avoid, although I am not sure why, i just found [this note](https://github.com/kulshekhar/ts-jest/blob/25e1c63dd3797793b0f46fa52fdee580b46f66ae/src/compiler.ts#L141-L142), where it says that it is slow:
> [...] If we don't return `undefined` it results in `undefined === "undefined"` and run `createProgram` again (which is very slow).

On the other hand `ts-jest` without `isolatedModules` became slower in bigger projects since v25.10, so maybe `createProgram` scales better? I am not a TypeScript internals expert, this is just a baseless guess...

`ts.createProgram` is required to gain access to the type checker, which is extensively used in the transformer, but is not available in the lightweight compiler `ts-jest` offers. Rewriting this transformer without the type checker is a task that is quite complicated, if not impossible (the type checker keeps track of class declarations in the project and can then re-reference them from other files).

I also used `@angular/compiler-cli` to parse the project configuration and the transformer itself, which is located inside `@ngtools/webpack`.

The wrapper would look something like this:
```js
const { readConfiguration } = require('@angular/compiler-cli')
const { downlevelConstructorParameters } = require('@ngtools/webpack/src/transformers/ctor-parameters.js')
const ts = require('typescript')

function factory(cs) {
  const config = readConfiguration(cs.tsJest.tsConfig.value)
  const program = ts.createProgram(config.rootNames, config.options || {}, config.compilerHost)

  return downlevelConstructorParameters(() => program.getTypeChecker())
}
```

I would require more feedback from you guys, @thymikee and especially @ahnpnl or someone else from `ts-jest`, who is familiar with the TypeScript internals going on in `ts-jest`.

**Comment by schuchard** [comment]
FWIW - I've updated the Briebug schematic to support Angular 8. It includes some of the changes mentioned above in version 2.0.0 - https://github.com/briebug/jest-schematic

**Comment by joebnb** [comment]
solved my problem,nice

**Comment by dalitroz** [comment]
The only thing that worked for me, after trying all the other options here,was to enable the following in the compilerOptions in the tsconfig.json file:
    "emitDecoratorMetadata": true,
This option should be enabled if the  "experimentalDecorators": true.
 I had to enable both in order to make my app work.
Hope that helps someone.

**Comment by ahnpnl** [comment]
> ## Update
> I figured out we can create the program with the ts compiler module inside `configSet`, which removes the direct dependency to `typescript` and `@angular/compiler-cli`:
> 
> ```js
> const { downlevelConstructorParameters } = require('@ngtools/webpack/src/transformers/ctor-parameters.js')
> 
> function factory(cs) {
>   const program = cs.compilerModule.createProgram(cs.typescript.fileNames, cs.typescript.options, undefined)
>   return downlevelConstructorParameters(() => program.getTypeChecker())
> }
> ```
> 
> If the TS Program and Type Checker would be instantiated by `ts-jest` and available to the transformer, we could just include the original Angular transformer. See [kulshekhar/ts-jest#1146](https://github.com/kulshekhar/ts-jest/issues/1146) for the this request.

@who I think we can start working on this right ?, based on these codes from you + the exposed Program from ts-jest. 

Currently isolatedModules mode doesn’t expose Program so I don’t know if it is a must to have or just use compileModule.createProgram as fallback for isolatedModules mode like you did above ?

**Comment by wtho** [comment]
@ahnpnl yes, I have a working implementation for `isolatedModules: false` using the [original Angular Transformer](https://github.com/angular/angular-cli/blob/1f1600225508ba58d2dfa617747cd9ffd4e31c2e/packages/ngtools/webpack/src/transformers/ctor-parameters.ts). I am not sure what the better alternative would be:
* recommending the workaround
* discourage from using `isolatedModules: true`

Finally the best solution for everybody would be a performant `ts-jest` without isolated modules, but it is difficult ot estimate how difficult it is to achieve this, or if it can be achieved at all.

**Comment by ahnpnl** [comment]
Hmm actually I gonna create a PR to expose `Program` also for `isolatedModules: true` because the need of support `projectReferences` which also requires `Program`. So no matter what kind of settings for `isolatedModules`, there will be always a `Program` instance. 

I noticed that the impact to performance is usually noticeable when using `Program.emit`, otherwise creating a `Program` instance doesn't do anything. I hope I'm right. Do you have similar observation ?

`LanguageService` in `ts-jest` has been improved quite a lot but not sure what else it can be done. There will be room for improvement but will need some times to figure out and helps from others.

**Comment by wtho** [comment]
Ok, awesome! This simplifies things for this issue.

Will prepare the PR soon!

**Comment by norato** [comment]
The error
`This constructor is not compatible with Angular Dependency Injection because its dependency at index 0 of the parameter list is invalid.`

is fixed using this `emitDecoratorMetadata: true`

**Comment by wtho** [comment]
@norato this workaround was mentioned several times before in this thread, but is just a workaround and not more, not a fix for the preset.

If you wonder why this is not closed: The fix in this preset would not require any changes in the `tsconfig`, as this preset in a perfect world would already include all required settings.

**Comment by norato** [comment]
@wtho I did not found those references mentioned when I wrote this comment. This was just to help the ones that need to "fix" this error until we don't have the solution.

**Comment by villelahdenvuo** [comment]
This is a weird one. I'm also getting this error, however, I do have `emitDecoratorMetadata: true` in my tsconfig.spec.json. But the problem only seems to appear with a specific provider that doesn't want to work unless I reference it in the file outside of type definitions.

**Comment by wtho** [comment]
@villelahdenvuo `isolatedModules` switches off typechecking and just compiles TS to JS.

Do you have an own file for each injectable class?

**Comment by villelahdenvuo** [comment]
@wtho Yeah, but I actually re-export the class through a another file, that might be why it's failing.

**Comment by villelahdenvuo** [comment]
@wtho I can confirm that moving the failing class to the file being imported made the provider work again even without isolatedModules.

**Comment by Murkhassan86** [comment]
![image](https://user-images.githubusercontent.com/67417764/85708492-15c98000-b699-11ea-8ecf-b7fdefb57766.png)
[](url)
I have created component with name 'test' and after simple coding i am facing this type of error. I have also make this true "emitDecoratorMetadata": true,. still i am getting this error.
I am using angular framework and recently joined GitHub. kindly help me out.

**Comment by Murkhassan86** [comment]
![image](https://user-images.githubusercontent.com/67417764/85709303-fbdc6d00-b699-11ea-9d9b-ceb3a7f40b04.png)
this is how my test.component.ts page looks like. In test.component.html i have created simple form by using some input fields only. what is the error in this code. your response will highly be appreciated.

**Comment by wtho** [comment]
Hey @Murkhassan86,
you definitely have a different issue, you seem to not have worked much with TypeScript syntax yet.

Try this instead:
```ts
constructor(private fb: FormBuilder) {}
```

Cheers!

**Comment by Murkhassan86** [comment]
Dear wtho, 
thank you for the correction.

**Comment by senseihimanshu** [comment]
Why I'm getting this error? "emitDecoratorMetadata": true, doesn't work.

```
import { Component, OnInit, Input, ChangeDetectorRef, Attribute, Output, EventEmitter } from '@angular/core';
import { FormGroup, FormControl } from '@angular/forms';
import { Store } from '@ngrx/store';
import { takeWhile } from 'rxjs/operators';

@Component({
  selector: 'keyword-list',
  templateUrl: './keyword-list.component.html',
  styleUrls: ['./keyword-list.component.css']
})
export class KeywordListComponent implements OnInit {

  @Input() keywordForm: FormGroup;
  @Input() preSelectedKeywords: any;
  @Input() singleSelect: boolean = false;
  @Input() isOptional: boolean = false;
  @Output() filedSelected = new EventEmitter();
  public keywordList: any[];
  public selectedKeywords: any[];
  public loading: boolean;
  showWarIcon: boolean = false;
  private isActive: boolean = true;
  private isAddedonoutFocus: boolean = false;
  @Input() heading: string = null;
  @Input() placeholder: string = null;  

  public placeholderText: string;
  constructor(private masterDataStore: any,
    private changeDetectorRef: ChangeDetectorRef,
    @Attribute('isSkills') public isSkills: boolean = false) {
    this.selectedKeywords = [];
    this.loading = false;
  }

  ngOnInit() {
    this.placeholderText = this.placeholder;
    this.setPreSelectedKeywords();
  }

  ngOnDestroy() {
    this.isActive = false;
  }

  public clearSearchResult(event: any): void {
    this.keywordList = [];
  }

  public selectKeyword(item: any): void {
    let selectedItem = { id: item, text: item };
    if (this.singleSelect)
      this.selectedKeywords = [selectedItem];
    else
      this.selectedKeywords.push(selectedItem);
    (<FormControl>this.keywordForm.controls["keywords"]).setValue('', { onlySelf: false });
    this.keywordList = [];
    this.filedSelected.emit();
    this.placeholderText = '';
  }

  public changeKeyword(): void {
    if (this.keywordForm.value.keywords !== undefined && this.keywordForm.value.keywords !== "" && this.keywordForm.value.keywords.includes(',')) {
      if (this.keywordForm.value.keywords.length > 1) {
        let prekeywords = this.selectedKeywords;
        let item = this.keywordForm.value.keywords.split(',')[0];
        if (item && item !== '') {
          let selectedItem = { id: item, text: item };
          var keyword = this.selectedKeywords.filter(x => x.text == selectedItem.text)[0];
          var index = this.selectedKeywords.indexOf(keyword);
          if (!(index >= 0)) {
            prekeywords.push(selectedItem);
          }
          if (this.singleSelect)
            this.selectedKeywords = [selectedItem];
          else {
            this.selectedKeywords = prekeywords;
          }
        }
      }

      (<FormControl>this.keywordForm.controls["keywords"]).setValue('', { onlySelf: false });
      this.keywordList = [];
    }
    else if (this.keywordForm.value.keywords !== undefined && this.keywordForm.value.keywords !== "") {
      // this.masterDataStore.dispatch(new MasterData_LoadKeywords(this.keywordForm.value.keywords, false));
    } else {
      this.keywordList = [];
    }
  }

  public focusOutFunction(): void {

    (<FormControl>this.keywordForm.controls["keywords"]).setValue('', { onlySelf: false });
    // let value = this.keywordForm.get["keywords"].value;
    // if(value){
    this.filedSelected.emit();
    // }
  }

  public removeKeyword(item: any): void {
    this.selectedKeywords.splice(this.selectedKeywords.indexOf(item), 1);
    if (!this.selectedKeywords || this.selectedKeywords.length == 0) {
      this.placeholderText = this.placeholder;
    }
  }

  private setPreSelectedKeywords() {
    if (this.preSelectedKeywords && this.preSelectedKeywords.length > 0) {
      this.placeholderText = '';
      let keywords = [];
      this.preSelectedKeywords.forEach(x => {
        if (x)
          keywords.push({ id: x, text: x });
      });
      if (keywords.length > 0) {
        this.selectedKeywords.push(...keywords);
      }
    }
  }
}

```

**Comment by dannyskoog** [comment]
@senseihimanshu Most likely due to that you're declaring `masterDataStore` as `any`. You need to provide a proper (and registered) type in order for Angular's DI system to know what instance to inject

**Comment by systemdice** [comment]
> ### TL;DR
> ```diff
> tsconfig.spec.json
> 
>    "compilerOptions": {
> +    "emitDecoratorMetadata": true,
>      "outDir": "./out-tsc/spec",
> ```
> 
> ### Issue
> The Angular CLI takes over some of the TS -> JS transpilation using transformers, especially when it's about the decorators and reflection. We were not aware of this change, but it appearently came in in Angular v8.1.
> 
> Here the full PR: [angular/angular-cli#14473](https://github.com/angular/angular-cli/pull/14473)
> 
> We can try to reuse their transformer (https://github.com/angular/angular-cli/blob/master/packages/ngtools/webpack/src/transformers/ctor-parameters.ts), as `@angular/cli` is anyway a dependency in every Angular project.
> 
> I will try to experiment with the transformer, we can track the progress in this issue till then.
> 
> ### Workaround
> To make `ts-jest` not lose the parameter information upon compilation without the AST transformation, you can set `"emitDecoratorMetadata": true` in your `tsconfig.spec.json` for now. The PR mentions issues with AOT compilation, which is usually not used in testing, so this setting should not create any problems. If you use karma as well and this leads to issues, try to use a separate `tsconfig.spec.json`.

**Comment by systemdice** [comment]
You are awesome. It solved my issue. Iwas struggling for days and hours.

**Comment by federicobenedetti** [comment]
Hey there! Any news on this bug?
